File: remoter/src/main/java/remoter/compiler/builder/MapParamBuilder.java
package remoter.compiler.builder;

import com.squareup.javapoet.MethodSpec;

import java.util.HashMap;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;

/**
 * A {@link ParamBuilder} for Map type parameters
 */
class MapParamBuilder extends ParamBuilder {


    protected MapParamBuilder(Messager messager, Element element) {
        super(messager, element);
    }


    @Override
    public void writeParamsToProxy(VariableElement param, ParamType paramType, MethodSpec.Builder methodBuilder) {
        if (param.asType().getKind() == TypeKind.ARRAY) {
            logError("Map[] is not supported");
        } else {
            if (paramType != ParamType.OUT) {
                methodBuilder.addStatement("data.writeMap(" + param.getSimpleName() + ")");
            }
        }
    }

    @Override
    public void readResultsFromStub(TypeMirror resultType, MethodSpec.Builder methodBuilder) {
        if (resultType.getKind() == TypeKind.ARRAY) {
            logError("Map[] is not supported");
        } else {
            methodBuilder.addStatement("reply.writeMap(result)");
        }
    }

    @Override
    public void readOutResultsFromStub(VariableElement param, ParamType paramType, String paramName, MethodSpec.Builder methodBuilder) {
        methodBuilder.addStatement("reply.writeMap(" + paramName + ")");
    }


    @Override
    public void readResultsFromProxy(TypeMirror resultType, MethodSpec.Builder methodBuilder) {
        if (resultType.getKind() == TypeKind.ARRAY) {
            logError("Map[] is not supported");
        } else {
            methodBuilder.addStatement("result = reply.readHashMap(getClass().getClassLoader())");
        }
    }

    @Override
    public void writeParamsToStub(VariableElement param, ParamType paramType, String paramName, MethodSpec.Builder methodBuilder) {
        super.writeParamsToStub(param, paramType, paramName, methodBuilder);
        if (param.asType().getKind() == TypeKind.ARRAY) {
            logError("Map[] is not supported");
        } else {
            if (paramType == ParamType.OUT) {
                methodBuilder.addStatement(paramName + " = new $T()", HashMap.class);
            } else {
                methodBuilder.addStatement(paramName + " = data.readHashMap(getClass().getClassLoader())");
            }
        }
    }

    @Override
    public void readOutParamsFromProxy(VariableElement param, ParamType paramType, MethodSpec.Builder methodBuilder) {
        if (paramType != ParamType.IN) {
            methodBuilder.beginControlFlow("if (" + param.getSimpleName() +" != null)");
            methodBuilder.addStatement(param.getSimpleName() +".clear()");
            methodBuilder.addStatement("reply.readMap(" + param.getSimpleName() + ", getClass().getClassLoader())");
            methodBuilder.endControlFlow();
        }
    }
}

File: remoter/src/main/java/remoter/compiler/kbuilder/MapParamBuilder.kt
package remoter.compiler.kbuilder

import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.MAP
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.lang.model.element.VariableElement
import javax.lang.model.type.TypeKind
import javax.lang.model.type.TypeMirror


/**
 * A [ParamBuilder] for Map type parameters
 */
internal class MapParamBuilder(remoterInterfaceElement: Element, bindingManager: KBindingManager) : ParamBuilder(remoterInterfaceElement, bindingManager) {
    override fun writeParamsToProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {
        if (param.asType().kind == TypeKind.ARRAY) {
            logError("Map[] is not supported")
        } else {
            if (paramType != ParamType.OUT) {
                methodBuilder.addStatement("$DATA.writeMap(" + param.simpleName + " as %T<*, *>?)", MAP)
            }
        }
    }

    override fun readResultsFromStub(methodElement: ExecutableElement, resultType: TypeMirror, methodBuilder: FunSpec.Builder) {
        if (resultType.kind == TypeKind.ARRAY) {
            logError("Map[] is not supported")
        } else {
            methodBuilder.addStatement("$REPLY.writeMap($RESULT as %T<*, *>?)", MAP)
        }
    }

    override fun readOutResultsFromStub(param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {
        methodBuilder.addStatement("$REPLY.writeMap($paramName as %T<*, *>?)", MAP)
    }

    override fun readResultsFromProxy(methodType: ExecutableElement, methodBuilder: FunSpec.Builder) {
        val resultType = methodType.getReturnAsKotlinType()
        val resultMirror = methodType.getReturnAsTypeMirror()
        if (resultMirror.kind == TypeKind.ARRAY) {
            logError("Map[] is not supported")
        } else {
            methodBuilder.addStatement("$RESULT = $REPLY.readHashMap(javaClass.getClassLoader()) as %T", resultType)
        }
    }

    override fun writeParamsToStub(methodType: ExecutableElement, param: VariableElement, paramType: ParamType, paramName: String, methodBuilder: FunSpec.Builder) {
        super.writeParamsToStub(methodType, param, paramType, paramName, methodBuilder)
        if (param.asType().kind == TypeKind.ARRAY) {
            logError("Map[] is not supported")
        } else {
            if (paramType == ParamType.OUT) {
                methodBuilder.addStatement("$paramName = mutableMapOf()")
            } else {
                methodBuilder.addStatement("$paramName = $DATA.readHashMap(javaClass.getClassLoader()) as %T", param.asKotlinType())
            }
        }
    }

    override fun readOutParamsFromProxy(param: VariableElement, paramType: ParamType, methodBuilder: FunSpec.Builder) {
        if (paramType != ParamType.IN) {
            if (param.isNullable()){
                methodBuilder.beginControlFlow("if (${param.simpleName} != null)")
            }

            methodBuilder.addStatement( "${param.simpleName}.clear()")
            methodBuilder.addStatement("$REPLY.readMap(" + param.simpleName + " as %T<*, *>, javaClass.getClassLoader())", MAP)

            if (param.isNullable()){
                methodBuilder.endControlFlow()
            }
        }
    }
}
